src/PAModel/Serializers/SourceSerializer.cs

```diff
@@ -795,5 +798,62 @@ private static BuildVerJson GetBuildDetails()
             }
         }
 
+        /// <summary>
+        /// This method validates if the templates being references in the sources do exist.
+        /// </summary>
+        private static void ValidateIfTemplateExists(CanvasDocument app, BlockNode node, BlockNode root, ErrorContainer errors)
+        {
+            foreach (var child in node.Children)
+            {
+                // group, grouContainer, gallery etc. have nested controls so run the validation for all the children.
+                if (child.Children?.Count > 0)
+                {
+                    foreach (var child1 in child.Children)
+                    {
+                        ValidateIfTemplateExists(app, child1, root, errors);
+                    }
+                }
+
+                CombinedTemplateState templateState;
+                app._templateStore.TryGetTemplate(child.Name.Kind.TypeName, out templateState);
+
+                // Some of the child components don't have a template eg. TestStep, so we can safely continue if we can't find an entry in the templateStore.
+                if (templateState == null)
+                {
+                    continue;
+                }
+
+                // If its a widget template then there must be a xml file in the pkgs directory.
+                if (templateState.IsWidgetTemplate)
+                {
+                    if (!app._templates.UsedTemplates.Any(x => x.Name == child.Name.Kind.TypeName))
+                    {
+                        errors.ValidationError(root.SourceSpan.GetValueOrDefault(), $"Widget control template: {templateState.Name}, version {templateState.Version} was not found in the pkgs directory and is referred in {root.Name.Identifier}. " +
```

Comment:
>.GetValueOrDefault() [](http://example.com/codeflow?start=62&length=20)

Can omit this.

Reply:
the AddError takes in "SourceLocation" where the root.SourceSpan is of nullable type "SoruceLocation?", so either I can do the default handling here (like its currently done) or can do that in the ValidateError overload. This seems to be a better palce to handle.